sklearn/utils/metaestimators.py

```diff
@@ -82,53 +82,89 @@ def _validate_names(self, names):
         # update the docstring of the descriptor
         update_wrapper(self, fn)
 
-    def __get__(self, obj, type=None):
-        # raise an AttributeError if the attribute is not present on the object
+    def __get__(self, obj, owner=None):
         if obj is not None:
             # delegate only on instances, not the classes.
             # this is to allow access to the docstrings.
-            for delegate_name in self.delegate_names:
-                try:
-                    delegate = attrgetter(delegate_name)(obj)
-                except AttributeError:
-                    continue
-                else:
-                    getattr(delegate, self.attribute_name)
-                    break
-            else:
-                attrgetter(self.delegate_names[-1])(obj)
+            if not self.check(obj):
+                raise AttributeError(
+                    f"This {repr(owner.__name__)}"
+                    f" has no attribute"
+                    f" {repr(self.attribute_name)}"
+                )
 
         # lambda, but not partial, allows help() to work with update_wrapper
-        out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)
+        out = lambda *args, **kwargs: self.fn(obj, *args, **kwargs)  # noqa
         # update the docstring of the returned function
         update_wrapper(out, self.fn)
         return out
 
 
+def available_if(check):
+    """An attribute that is available only if check returns a truthy value
+
+    Parameters
+    ----------
+    check : callable
+        When passed the object with the decorated method, this should return
+        a truthy value if the attribute is available, and either return False
+        or raise an AttributeError if not available.
+    """
+    return lambda fn: _AvailableIfDescriptor(fn, check, attribute_name=fn.__name__)
+
+
+class _IffHasAttrDescriptor(_AvailableIfDescriptor):
+    """Implements a conditional property using the descriptor protocol.
+
+    Using this class to create a decorator will raise an ``AttributeError``
+    if none of the delegates (specified in ``delegate_names``) is an attribute
+    of the base object or the first found delegate does not have an attribute
+    ``attribute_name``.
+
+    This allows ducktyping of the decorated method based on
+    ``delegate.attribute_name``. Here ``delegate`` is the first item in
+    ``delegate_names`` for which ``hasattr(object, delegate) is True``.
+
+    See https://docs.python.org/3/howto/descriptor.html for an explanation of
+    descriptors.
+    """
+
+    def __init__(self, fn, delegate_names, attribute_name):
+        super().__init__(fn, self._check, attribute_name)
+        self.delegate_names = delegate_names
+
+    def _check(self, obj):
+        for delegate_name in self.delegate_names:
+            try:
+                delegate = attrgetter(delegate_name)(obj)
+            except AttributeError:
+                continue
+            break
+        else:
+            return False
```

Comment:
Do you think using a sentinel value is more readable? Something like this:

```python
delegate = None
for delegate_name in self.delegate_names:
    try:
        delegate = attrgetter(delegate_names)(obj)
        break
    except AttributeError:
        continue
        
if delegate is None:
    return False

return getattr(delegate, self.attribute_name) or True
```

Reply:
Okay... not too fussed